src/Middleware/Diagnostics.EntityFrameworkCore/src/HttpContextDatabaseContextDetailsExtensions.cs

```diff
@@ -0,0 +1,87 @@
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Metadata.Conventions;
+using Microsoft.EntityFrameworkCore.Metadata.Conventions.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore
+{
+    internal static class HttpContextDatabaseContextDetailsExtensions
+    {
+        public static async ValueTask<DatabaseContextDetails?> GetContextDetailsAsync(this HttpContext httpContext, Type dbcontextType, ILogger logger)
+        {
+            var context = (DbContext?)httpContext.RequestServices.GetService(dbcontextType);
+
+            if (context == null)
+            {
+                logger.ContextNotRegisteredDatabaseErrorPageMiddleware(dbcontextType.FullName);
+                return null;
+            }
+
+            var relationalDatabaseCreator = context.GetService<IDatabaseCreator>() as IRelationalDatabaseCreator;
+            if (relationalDatabaseCreator == null)
+            {
+                logger.NotRelationalDatabase();
+                return null;
+            }
+
+            var databaseExists = await relationalDatabaseCreator.ExistsAsync();
+
+            if (databaseExists)
+            {
+                databaseExists = await relationalDatabaseCreator.HasTablesAsync();
+            }
+
+            var migrationsAssembly = context.GetService<IMigrationsAssembly>();
+            var modelDiffer = context.GetService<IMigrationsModelDiffer>();
+
+            var snapshotModel = migrationsAssembly.ModelSnapshot?.Model;
+            if (snapshotModel is IConventionModel conventionModel)
+            {
+                var conventionSet = context.GetService<IConventionSetBuilder>().CreateConventionSet();
+
+                var typeMappingConvention = conventionSet.ModelFinalizingConventions.OfType<TypeMappingConvention>().FirstOrDefault();
+                if (typeMappingConvention != null)
+                {
+                    typeMappingConvention.ProcessModelFinalizing(conventionModel.Builder, null);
+                }
+
+                var relationalModelConvention = conventionSet.ModelFinalizedConventions.OfType<RelationalModelConvention>().FirstOrDefault();
+                if (relationalModelConvention != null)
+                {
+                    snapshotModel = relationalModelConvention.ProcessModelFinalized(conventionModel);
+                }
+            }
+
+            if (snapshotModel is IMutableModel mutableModel)
+            {
+                snapshotModel = mutableModel.FinalizeModel();
+            }
```

Comment:
@AndriySvyryd Is all this still necessary? Is there an easier way?

Reply:
Yes. Easier way will be enabled by https://github.com/dotnet/efcore/issues/22031

Reply:
Technically we should also call `SnapshotModelProcessor.Process()` to fix old snapshots, but this would bring in all kinds of dependencies